mmv1/products/certificatemanager/terraform.yaml

```diff
@@ -44,14 +44,20 @@ overrides: !ruby/object:Overrides::ResourceOverrides
           cert_name: "dns-cert"
         ignore_read_extra:
           - "managed.0.dns_authorizations"
+    examples:
+      - !ruby/object:Provider::Terraform::Examples
+        name: "certificate_manager_self_managed_certificate"
+        primary_resource_id: "default"
+        vars:
+          cert_name: "self-managed-cert"
     properties:
       managed.dnsAuthorizations: !ruby/object:Overrides::Terraform::PropertyOverride
         # We don't support ignore_read on nested fields
         ignore_read: true
         custom_flatten: "templates/terraform/custom_flatten/certificate_manager_certificate_managed_dns_auth.go.erb"
-      selfManaged.certificatePem: !ruby/object:Overrides::Terraform::PropertyOverride
+      selfManaged.pemCertificate: !ruby/object:Overrides::Terraform::PropertyOverride
```

Comment:
it looks like both of these fields are input only, so normally we would add
`ignore_read: true`, however, looking 3 lines before this it looks like that doesn't work on nested fields, so we'll probably just want to copy how `managed.dnsAuthorizations` works by adding a `custom_flatten` for each. This is for both `selfManaged.pemCertificate` and `selfManaged.pemPrivateKey`

You'll also want to add `ignore_read_extra` on your new test and include both of these fields, in the same way the managed certificate test does with `managed.0.dns_authorizations` Does that make sense?

Reply:
I've added `ignore_read: true` and `ignore_read_extra`. I assume it doesn't make sense to `add custom_flatten` function - these PEM fields are not arrays, just regular strings. There is nothing to flatten.

Let me know if there's anything else to fix.

Reply:
@pawel-grz-krawczyk I understand the confusion. The way that magic modules works is that it creates a flatten function for all fields downstream. For example, your current code will generate [this snippet](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6447-old..auto-pr-6447#diff-d3c6d2e8722387ee9af946581a90dcfb53503f317325d924df142a6cedcaa832R521), which basically says "write to state what was returned from the API", but the API returns nothing, so we're storing nothing in state. Thus, when it's time to run the next plan, it shows a diff.

          + self_managed { # forces replacement
                + pem_certificate = (sensitive value)
                + pem_private_key = (sensitive value)
              }
          }

If you look at the [custom flatten function for `managed.dnsAuthorizations`](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/mmv1/templates/terraform/custom_flatten/certificate_manager_certificate_managed_dns_auth.go.erb), you'll see that what it is doing instead is that it will return whatever is in the config, since nothing is returned from the API. I think we'll want to do the same thing here. So you'll need to add 2 new files to `templates/terraform/custom_flatten`, one for each field, and basically just read what's in the config instead of what's returned from the API. Does that help?

Reply:
That explains a lot. These flatten functions confused me a lot. Especially given the zero right in the middle of `"managed.0.dns_authorizations"`. Seeing this zero I subconsciously assumed that it's like `"managed.dns_authorizations[0]"` - taking the first element from the array for some strange reason.

Thank you for the explanation, much appreciated! I'll add custom flatten for both PEM fields.

Reply:
I assume that I still need to add this zero in the middle in the flatten function:

```
return d.Get("self_managed.0.pem_private_key")
...
return d.Get("self_managed.0.pem_certificate")
```

Do I guess correctly that this zero is needed because for some reason fields in `exactly_one_of:` are "represented" as arrays (od size 0 or 1)? That's why when getting the value from `d *schema.ResourceData` you need to specify the full path of the field you want to get, including the index of an array. Is that intuition correct?

Reply:
Something is still wrong :( see logs [TestAccCertificateManagerCertificate_certificateManagerSelfManagedCertificateExample](https://github.com/GoogleCloudPlatform/magic-modules/pull/6447#issuecomment-1231457122):

```
    provider_test.go:315: Step 1/2 error: After applying this test step, the plan was not empty.
        stdout:
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
        -/+ destroy and then create replacement
        
        Terraform will perform the following actions:
        
          # google_certificate_manager_certificate.default must be replaced
        -/+ resource "google_certificate_manager_certificate" "default" {
              ~ id          = "projects/ci-test-project-188019/locations/global/certificates/tf-test-self-managed-certhc6gwdoblg" -> (known after apply)
                name        = "tf-test-self-managed-certhc6gwdoblg"
              ~ project     = "ci-test-project-188019" -> (known after apply)
                # (2 unchanged attributes hidden)
        
              + self_managed { # forces replacement
                  + pem_certificate = (sensitive value)
                  + pem_private_key = (sensitive value)
                }
            }
        
        Plan: 1 to add, 0 to change, 1 to destroy.
```

Reply:
Ok! I dove into this a bit and I think we also need to address the `selfManaged` flatten function. It turns out that the `selfManaged` block is returned with nothing in it - it then hits [this line](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-6447-old..auto-pr-6447#diff-d3c6d2e8722387ee9af946581a90dcfb53503f317325d924df142a6cedcaa832R511) and doesn't actually call the flatten functions for the `pemCertificate` or `pemPrivateKey`. It seems a bit like overkill, but I think we'll need to add a custom flatten function in there as well. Feel free to find a better way to do it! I was just playing with it while trying to diagnose the problem, and this was the easiest way to solve it. I just removed that check if len was 0.

Reply:
I've added custom flatten function for the self_managed field.

I have a couple of questions though. See other comments.